js_modules/dagit/packages/ui/src/components/ProductTour.tsx

```diff
@@ -0,0 +1,230 @@
+import React from 'react';
+import {createPortal} from 'react-dom';
+import styled, {CSSProperties} from 'styled-components/macro';
+
+import {Box} from './Box';
+import {Button} from './Button';
+import {Colors} from './Colors';
+import {Subheading} from './Text';
+
+export enum Position {
+  TOP_LEFT,
+  TOP_CENTER,
+  TOP_RIGHT,
+  BOTTOM_LEFT,
+  BOTTOM_CENTER,
+  BOTTOM_RIGHT,
+}
+
+type ObjectType =
+  | {img: string; video?: undefined; object?: undefined}
+  | {video: string; img?: undefined; object?: undefined}
+  | {object: React.ReactNode; video?: undefined; img?: undefined}
+  | {img?: undefined; video?: undefined; object?: undefined};
```

Comment:
This is kind of a confusing object type. Could it be a tagged union, and you refine on the `type` field?

```
type ObjectType =
| {type: 'image', img: string;}
| {type: 'video', video: string;}
| {type: 'object', object: React.ReactNode;}
| {type: 'empty'}
```

Reply:
You would have to pass type in as a prop and I think thats awkward from a user perspective.

From a user perspective this type definition is the friendliest imo since the call site doesnt have to do anything special.

Reply:
Well, I was thinking it would be a prop like `thing={{type: 'image', img: 'https://...'}}`.

I don't feel that strongly about it, it just looked a little unusual to me. :)

Reply:
Yeah I don't like having to pass a type explicitly. If we were to pass the type explicitly there would be no need to use different keys for image/video/object also. I'm essentially allowing the key you use to encode the type.